### DISCUSSION TITLE
Tana GPT-3 integration

### DISCUSSION THREAD

**Discussion by foeken** [johnlindquist/kit#954] [Share]
This script allows for Tana to call GPT-3 and return the results in the Tana paste format, you can also add native fields fields below the copied node to act as parameters for the GPT-3 query

<img width="929" alt="CleanShot 2022-11-28 at 17 46 29@2x" src="https://user-images.githubusercontent.com/13864/204334452-7e308d3d-0a2d-4c57-bead-44d864f57530.png">

[Open openai-replace in Script Kit](https://scriptkit.com/api/new?name=openai-replace&url=https://gist.githubusercontent.com/foeken/d43c929d5b27cd38ef694f17604b3fa5/raw/60d3841f721c2f990a9e7113f01018e8628269f9/openai-replace.js")

```js
// Name: OpenAI Replace
// Description: Replace using Open AI's API
// Shortcut: cmd ctrl s

import "@johnlindquist/kit"

let { Configuration, OpenAIApi } = await npm("openai")

let configuration = new Configuration({
  apiKey: await env("OPENAI_API_KEY"),
})

let openai = new OpenAIApi(configuration)
let complete = async (prompt, maxTokens, temperature) => {
    setTimeout(() => {
        setLoading(true)
    }, 250)
  console.log(maxTokens)
  console.log(temperature)
    console.log(prompt)
    let response = await openai.createCompletion({
        model: "text-davinci-003",
        prompt: `${prompt}`,
        temperature: temperature,
        max_tokens: maxTokens,
        top_p: 1,
        frequency_penalty: 0,
        presence_penalty: 0,
    })
    setLoading(false)
    return response?.data?.choices[0]?.text?.trim()  
}

let prompt = await getSelectedText()

let regexMatch = null

let temperature = 0.7
let temperatureRegex = /\n  Temperature: ([+-]?([0-9]*[.])?[0-9]+)/

regexMatch = temperatureRegex.exec(prompt)
if (regexMatch) {    
  temperature = parseFloat(regexMatch[1])
  prompt = prompt.replace(temperatureRegex, "")
}

let maxTokens = 512
let maxTokensRegex = /\n  Max Tokens: ([+-]?([0-9]*[.])?[0-9]+)/

regexMatch = maxTokensRegex.exec(prompt)
if (regexMatch) {    
  maxTokens = parseInt(regexMatch[1])
  prompt = prompt.replace(maxTokensRegex, "")
}

let result = ""

while (true) {
  if (result != "") {
    prompt = await editor(prompt) 
  }  

  div("Give me a second ...", "text-center text-4xl p-10 font-semibold")

  result = await complete(prompt, maxTokens, temperature)

  let retryOrContinue = await arg(
    {
      placeholder: "",
      hint: "[R]etry another prompt, [E]dit the reply, or continue (enter)",
      ignoreBlur: true,
    }, div(md(result), 'm-4')
  )        

    if (retryOrContinue === "E") {
        result = await editor(result)
        break
    }

    if (retryOrContinue === "") break
}

setSelectedText("%%tana%%\n" + result)

```

**Comment by wahidyahya** [comment]
Cool 👍

**Comment by arvindhans** [comment]
Hi Foeken! This is awesome! Could you please let me know how to install / import this script to Tana? I don't know much about code.

**Reply by wahidyahya** [reply]
I, too,  would some help with that.

**Reply by Hileotech** [reply]
If I got it right you simply have to install Scriptkit on your Mac (Windows version is still in Alpha at the moment I write) and then paste the code in ScriptKit (or follow the "Open openai-replace in Script Kit" link above the code).
I'm curious too, but as I'm totally new to Tana I would prefer to learn the core and then to "expand" Tana possibilities, so I think I'll wait for some days before digging into AI+Tana

**Reply by foeken** [reply]
Correct. Download ScriptKit and install this script either through the community scrips section on the website or by copy/paste, assign keyboard shortcut and done!

**Reply by wahidyahya** [reply]
Downloaded and installed Scriptkit
Installed this script 
Changed the shortcut 
Ran it 
Here is what I got "Request failed with status code 401."
I am new to this. 
So I guess I am doing something wrong.

**Reply by xgorse** [reply]
@wahidyahya you must add you're own OpenAI API key in the script 
```
let configuration = new Configuration({
  apiKey: await env("OPENAI_API_KEY"),
})
```
https://help.openai.com/en/articles/4936850-where-do-i-find-my-secret-api-key

**Reply by wahidyahya** [reply]
Thank you @xgorse,  I appreciate the tip. I believe I also work have to pay for the service for it to work.

**Comment by MrTownyp** [comment]
When I use it, it shows "openai is required for this script" in my mac and two options "Abort" & "Install openai". After I click  "Install openai" , it shows nothing. I don't know why

**Reply by johnlindquist** [reply]
@MrTownyp any chance you could make a screen recording of what "nothing happens" looks like?

**Reply by MrTownyp** [reply]
This is the situation.
https://user-images.githubusercontent.com/23568668/218297299-bfbe4e43-b5d5-4aa2-a64a-5836df7ae967.mp4
